File: R/summaries.R
# Summary helpers for prior draws

hdi_interval <- function(x, mass = 0.9) {
  x <- sort(x)
  n <- length(x)
  if (n == 0) {
    return(c(lower = NA_real_, upper = NA_real_))
  }
  m <- floor(mass * n)
  if (m < 1) {
    return(c(lower = NA_real_, upper = NA_real_))
  }
  widths <- x[(m + 1):n] - x[1:(n - m)]
  idx <- which.min(widths)
  c(lower = x[idx], upper = x[idx + m])
}

summarize_draws <- function(draws, hdi_mass = 0.9) {
  rows <- lapply(names(draws), function(name) {
    x <- draws[[name]]
    hdi <- hdi_interval(x, hdi_mass)
    qs <- stats::quantile(x, probs = c(0.1, 0.25, 0.5, 0.75, 0.9), names = FALSE)
    data.frame(
      parameter = name,
      mean = mean(x),
      sd = stats::sd(x),
      mad = stats::mad(x),
      hdi_lower = hdi[["lower"]],
      hdi_upper = hdi[["upper"]],
      p10 = qs[[1]],
      p25 = qs[[2]],
      p50 = qs[[3]],
      p75 = qs[[4]],
      p90 = qs[[5]],
      stringsAsFactors = FALSE
    )
  })
  do.call(rbind, rows)
}

File: web/js/summaries.js
// Summary statistics for prior draws. Mirrors R/summaries.R; quantiles use
// R's default type-7 definition and MAD uses R's 1.4826 consistency constant.

export function mean(x) {
  let total = 0;
  for (let i = 0; i < x.length; i++) total += x[i];
  return total / x.length;
}

export function sd(x) {
  const m = mean(x);
  let total = 0;
  for (let i = 0; i < x.length; i++) {
    const d = x[i] - m;
    total += d * d;
  }
  return Math.sqrt(total / (x.length - 1));
}

// R type-7 quantile on a pre-sorted array.
export function quantileSorted(sorted, p) {
  const n = sorted.length;
  if (n === 0) return NaN;
  const h = (n - 1) * p;
  const lo = Math.floor(h);
  const hi = Math.min(lo + 1, n - 1);
  return sorted[lo] + (h - lo) * (sorted[hi] - sorted[lo]);
}

export function quantile(x, p) {
  const sorted = Float64Array.from(x).sort();
  return quantileSorted(sorted, p);
}

export function median(x) {
  return quantile(x, 0.5);
}

// Median absolute deviation scaled by 1.4826 (consistent with sd for normals).
export function mad(x) {
  const m = median(x);
  const deviations = Float64Array.from(x, (v) => Math.abs(v - m));
  return 1.4826 * median(deviations);
}

// Narrowest interval containing `mass` of the draws (same algorithm as the
// Shiny app's hdi_interval).
export function hdiInterval(x, mass = 0.9) {
  const sorted = Float64Array.from(x).sort();
  const n = sorted.length;
  if (n === 0) return { lower: NaN, upper: NaN };
  const m = Math.floor(mass * n);
  if (m < 1 || m >= n) return { lower: NaN, upper: NaN };
  let bestIdx = 0;
  let bestWidth = Infinity;
  for (let i = 0; i + m < n; i++) {
    const width = sorted[i + m] - sorted[i];
    if (width < bestWidth) {
      bestWidth = width;
      bestIdx = i;
    }
  }
  return { lower: sorted[bestIdx], upper: sorted[bestIdx + m] };
}

export function summarizeDraws(draws, hdiMass = 0.9) {
  return Object.entries(draws).map(([parameter, x]) => {
    const sorted = Float64Array.from(x).sort();
    const hdi = hdiInterval(sorted, hdiMass);
    return {
      parameter,
      mean: mean(x),
      sd: sd(x),
      mad: mad(x),
      hdi_lower: hdi.lower,
      hdi_upper: hdi.upper,
      p10: quantileSorted(sorted, 0.1),
      p25: quantileSorted(sorted, 0.25),
      p50: quantileSorted(sorted, 0.5),
      p75: quantileSorted(sorted, 0.75),
      p90: quantileSorted(sorted, 0.9),
    };
  });
}

// Number formatting matching the Shiny app's summary table.
export function formatNumber(x) {
  if (x === null || x === undefined || Number.isNaN(x)) return "";
  if (x === 0) return "0";
  if (!Number.isFinite(x)) return x > 0 ? "Inf" : "-Inf";
  if (Math.abs(x) >= 1e6 || Math.abs(x) < 1e-4) {
    return x.toExponential(2);
  }
  return x.toFixed(3);
}
